src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "Trampoline.h"
+
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_split.h>
+#include <unistd.h>
+
+#include <cstdint>
+#include <string>
+
+#include "AllocateInTracee.h"
+#include "OrbitBase/ExecuteCommand.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+
+namespace orbit_user_space_instrumentation {
+
+using absl::numbers_internal::safe_strtou64_base;
+using orbit_base::ReadFileToString;
+
+bool DoAddressRangesOverlap(const AddressRange& a, const AddressRange& b) {
+  return !(b.end <= a.start || b.start >= a.end);
+}
+
+std::optional<size_t> LowestIntersectingAddressRange(const std::vector<AddressRange>& ranges_sorted,
+                                                     const AddressRange& range) {
+  for (size_t i = 0; i < ranges_sorted.size(); i++) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<size_t> HighestIntersectingAddressRange(
+    const std::vector<AddressRange>& ranges_sorted, const AddressRange& range) {
+  for (int i = ranges_sorted.size() - 1; i >= 0; i--) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+ErrorMessageOr<std::vector<AddressRange>> GetTakenAddressRanges(pid_t pid) {
```

Comment:
GetAllocated maybe?

Reply:
The range from zero to mmap_min_addr is not really allocated but just made unavailable via mmap for security reasons. The other ones are allocated (many of them but the loader I suppose). I agree that "Taken" is a bit hard to get. 
Changed to "GetUnavailableAddressRanges", agree?